### ISSUE TITLE
Propagate Standard Subset of Labels Across All K8s Repos (labels.yaml files need to be updated)

### ISSUE THREAD

**Issue by grodrigues3** [kubernetes/test-infra#2504]
kubernetes
kubernetes.github.io
contrib (might need manual renames first)
features

**Comment by grodrigues3** [comment]
We're going to try and use prow to keep track of when certain protected labels are changed in any repos.  If they are, create a new PR in the test-infra repo that modifies the source label file.

Better description here:
https://docs.google.com/a/google.com/document/d/1x373zsIWi69mbwIBjh-LIAofaw_cNt8TCB5NGUeUdEY/edit?usp=sharing

**Comment by grodrigues3** [comment]
@Sevii @lukaszgryglicki @dettervt 
All of you have been involved in contribX efforts and are looking for work.  This is pretty high priority, and if one of you could start working on the synchronization part, that'd be AWESOME.

1. Create a labels.yaml file in test infra repo that contains subset of labels we want in all repos (start with the area/, kind/ , priority/ , lgtm, approved and sig labels for now.
1. Write a prow plugin to check if someone creates or deletes one of the protected labels.  If that happens, recreate the label and direct them to change the yaml file (maybe open a PR and explain that the change will happen to ALL k8s repos)

Note: I'm not asking for the initial synchronization.  I think @spxtr was going to write some plumbing for that

**Comment by spxtr** [comment]
Writing a prow plugin is certainly possible, but it'll be tricky. One alternative is to write a program that runs daily under prow and does the following:

1. Get a list of all repos under the kubernetes org with [`GET /orgs/kubernetes/repos`](https://developer.github.com/v3/repos/#list-organization-repositories).
1. For each repo, list all labels in the repo with [`GET /repos/kubernetes/foo/labels`](https://developer.github.com/v3/issues/labels/#list-all-labels-for-this-repository).
1. If any labels are wrong, sync them, open issues, PRs, whatever. Hopefully this doesn't happen often.

This will be much simpler than trying to use a plugin. Each list API call can return up to a hundred entries per call. We could even run it hourly, since it probably won't even use a hundred API tokens per sync. The exception is the first time it is run. Then I think it will have to do `num_labels * num_repos` calls. We might do that with a one-off script, though.

**Comment by lukaszgryglicki** [comment]
I no one picks that, I can work on it after the Holidays.

**Comment by grodrigues3** [comment]
@lukaszgryglicki sorry, which holidays are you referring to?  I just want to get a sense of the timeline for this.

**Comment by dettervt** [comment]
If @lukaszgryglicki doesn't end up replying, I can take a look as well.

**Comment by lukaszgryglicki** [comment]
I was meaning Easter.
I'm back now, but I have other work to do currently, so @dettervt feel free to work on this.
I'll let You know when I have time to get back to kubernetes.

**Comment by dettervt** [comment]
Sure, I'll take a look. @grodrigues3 requested access to the linked doc

**Comment by lukaszgryglicki** [comment]
Also requested access - because it seems like I can work on it now.
In the meantime I'll review other issues on test-infra...
@dettervt are You working on this? Asking because I don't want to duplicate Your work?

**Comment by dettervt** [comment]
@lukasredynk I hadn't gotten a chance to start, all yours

**Comment by spxtr** [comment]
Note that the doc is a bit out of date, and I suggest following https://github.com/kubernetes/test-infra/issues/2504#issuecomment-294242096 rather than it.

**Comment by lukaszgryglicki** [comment]
OK so I will switch to this.

**Comment by lukaszgryglicki** [comment]
@spxtr I'll create program as You suggested.
Can You help me a little?
I have just a few questions:
-Can You point me to the place/files in prow where I can put this external program call?
-Are there any examples of such external program call from prow?
-Should it be written in golang (just like prow) or python, ruby etc. can be used too?
-One off script.... not exactly sure what is the difference. The new program will get all kubernetes repos, then all labels for each of them, and then sync. Basically this one off script would do exactly the same - the only difference would be that it will have to sync a lot more than usually.
-Prow already have access to kubernetes repos with some kind of OAuth2 etc, but this new tool will have to authorize externally to prow.... maybe really better to write a separate plugin that runs daily? If not then can You point me to some places where I can follow some standards to authorize such external program? The problem is that I'll probably use my own repo for testing locally, but final version will use some secrets from kubernetes repo - and then I have no idea if it will work or not.
-What exactly mean that labels are wrong... I can just see if some of them are missing or I can compare every label with labels.yaml (no case sensitive) and report different case or color for example. And lets assume we have 3 bad labels and 3 missing on one repo and similar situation in 2nd repo. Should I open one summary issue on 1st repo (with description what is wrong) and similar issue on 2nd repo? Or should I create separate issues for all wrong/missing labels on each repo?

Now I'm starting from 
-Getting a list of labels from kubernetes/kubernetes repo (labels from linked document) and create labels.yaml
-Read that file in external program (I'll start with totally idependent external program, I will then link to it prow eventually)
-Get list of repos, and then labels for it
-Check against labels.yaml, and aggregate wrong/missing labels per repo
-For each repo if there is at least 1 wrong/missing label - create an issue.

All this stuff will be done just externally to anything - we have to decide how exactly implement this.

**Comment by lukaszgryglicki** [comment]
I have most of the stuff in the above link...
-Configured labels.yaml
-Get all kubernetes repos (and stored locally in YAML file for local devel mode)
-Get all repos all labels (also stored in local YAML for local dev)
-Some first tests of YAML loading
-I will write sync function next (meaning function that detects missing labels on each repo).

I wonder if this program should update/add labels or just create issues - we need to decide what exactly it should do.
Creating issues seems ok but.... any next rund will create another issue unless I also implement some kind of checking if such an issue already exists.... and how to check that? Maybe that issue should have some special label like `labels-guardian` or anything?

Thoughts?

**Comment by grodrigues3** [comment]
cc @foxish

**Comment by spxtr** [comment]
> -Can You point me to the place/files in prow where I can put this external program call?
> -Are there any examples of such external program call from prow?

It can just run as a daily job on prow, which means you can use our standard job control flow. The code should probably live outside of the prow folder. One example of such a job is the triage job, [configured here](https://github.com/kubernetes/test-infra/blob/062d52d1043956f8a66480c410f75c7940f1bcd8/prow/config.yaml#L811-L826), [code here](https://github.com/kubernetes/test-infra/tree/master/triage).

> -Should it be written in golang (just like prow) or python, ruby etc. can be used too?

Doesn't matter as long as it works, but I suggest golang or python because I know those ones best.

> -One off script.... not exactly sure what is the difference. The new program will get all kubernetes repos, then all labels for each of them, and then sync. Basically this one off script would do exactly the same - the only difference would be that it will have to sync a lot more than usually.

Yeah, maybe it's not worth worrying about right now.

> -Prow already have access to kubernetes repos with some kind of OAuth2 etc, but this new tool will have to authorize externally to prow.... maybe really better to write a separate plugin that runs daily? If not then can You point me to some places where I can follow some standards to authorize such external program? The problem is that I'll probably use my own repo for testing locally, but final version will use some secrets from kubernetes repo - and then I have no idea if it will work or not.

The new tool will run in the prow cluster as a standard job, but it shouldn't assume access to the same k8s secret that prow uses. It might live in a different namespace, for instance. Instead we should create a secret in the cluster with a new GitHub API token. You can use go-github for easy communication with GitHub if you'd like. I think we already vendor it.

> -What exactly mean that labels are wrong... I can just see if some of them are missing or I can compare every label with labels.yaml (no case sensitive) and report different case or color for example. And lets assume we have 3 bad labels and 3 missing on one repo and similar situation in 2nd repo. Should I open one summary issue on 1st repo (with description what is wrong) and similar issue on 2nd repo? Or should I create separate issues for all wrong/missing labels on each repo?

To start out with I think we should just silently fix the labels without opening any issues. As you've noticed, there are some tricky edge cases involving opening issues that we want to get right, but at least for the first pass we don't want to open issues.

**Comment by lukaszgryglicki** [comment]
OK thanks.
For now I've started work in prow/label_sync.
I will finish it "as is" and then eventually adjust edge cases and details (like separate folder or separate secrets), for now they're shared with prow.
I will work on this probably the day after tomorrow, because I have something scheduled for tomorrow.
Or if IO manage to finish that scheduled task then I'll work on this as weel.
Can You take a very first look to give some critical feedback (I mean if I'm doing something REALLY bad) here:
https://github.com/kubernetes/test-infra/compare/master...lukaszgryglicki:lukaszgryglicki-2504-propagate-standard-labels

**Comment by spxtr** [comment]
It looks reasonable so far. I suggest moving it into `test-infra/label_sync`, since it's not a part of core prow but a prow job instead. If you'd like to use go-github so that you don't have to implement github API calls, feel free. Prow uses its own github client for various reasons, but you don't have to if you don't want to.

**Comment by lukaszgryglicki** [comment]
OK thanks, so I'll move it out of prow in final stage. For now I'lll finish it as is with full functionality, add test coverage etc.
In the meantime somebody should take a look on `labels.yaml`.
I've took those labels from `kubernetes/kubernetes` repo and filtered by doc provided:
```
sig/
kind/
area/
team/
priority/
lgtm
For-new-contributors
help-wanted
```

**Comment by lukaszgryglicki** [comment]
Moved code out of prow, I'll try to finish functionality needed and then think about configuring daily job for this - it seems like it need image with specific build - I'm not yet familiar with that process at all, so will have to follow some docs etc to create such images (speaking about new job in config.yaml).
OK working on syncing (I already have all data needed like loaded `labels.yaml` all kubernetes repos labels - now need a function that will return list of labels to be added and list of labels to be modified, then test coverage for this and finally GitHub calls that will actually do this).

**Comment by lukaszgryglicki** [comment]
`label_sync/main.go` added syncLabels function that gets required labels set and org's repo;'s labels set and returns UpdateData struct with operations required to sync all labels on all repos.

Added test suite for this function here:
`label_sync/main_test.go`

Will continue when I finish other tasks assigned on me (need to leave it for few days :( )
No PR yet, implementation here:
https://github.com/kubernetes/test-infra/compare/master...lukaszgryglicki:lukaszgryglicki-2504-propagate-standard-labels

**Comment by lukaszgryglicki** [comment]
Seems like I should have time to work on this task again tomorrow, will actually see.

**Comment by lukaszgryglicki** [comment]
Hi I've just finished program that syncs labels on all repos for given org (or for user:github_username).
I've used labels from kubernetes/kubernetes (requested in doc):
```
Protected Patterns
sig/*
kind/*
area/*
team/*
priority/*
Required Labels (See Kubernetes/Kubernetes for list)
lgtm
For-new-contributors
help-wanted
```

Labels are defined in `labels.yaml` file.

Here are results when I run this on my own repos (one of them synced):
https://github.com/lukaszgryglicki/store.js/labels
It handles request retrying in case of `first time` run with all labels missing etc. It successfully finished in about 10 minutes on ALL my public repos (like the one above) without any issue. And next run was 22s becasue it didn't needed to update anything (it only needed to check if all is OK).
I've also added detailed test coverage with edge cases in `label_sync/main_test.go`.

Now I ask for some help how to integrate it with out daily jobs etc.

Sources are here:
https://github.com/kubernetes/test-infra/compare/master...lukaszgryglicki:lukaszgryglicki-2504-propagate-standard-labels

It just synces all labels on all repos.
If label is missing - it adds that label.
If label have wrong name (mean wrong lower/upper case) it updates name to the name requested in labels.yaml (updates not deletes and adds, so labels will stay on issues, PRs etc.)
If label color is wrong it updates color.


Now I need a way of adding this as a daily job, I've seen @spxtr 's suggestion about adding this as a daily prow job - but I don't know how to create one - I see that it is probably the matter of generate entire OS image / pod with that tool installed - can anybody @spxtr @grodrigues3  help me doing this for a first time?

Also @spxtr @grodrigues3 can You take a look at final implementation?
This will change a little only because it needs to be connected as daily job - but I don't expect any logical changes there...
https://github.com/kubernetes/test-infra/compare/master...lukaszgryglicki:lukaszgryglicki-2504-propagate-standard-labels

Once I know how to connect this to prow daily job, I'll create PR in `kubernetes/test-infra` repo.

**Comment by grodrigues3** [comment]
I'll get to this later today.  Thanks for implementing!

**Comment by lukaszgryglicki** [comment]
Great, I will work on something other tomorrow (probably) and will get back to it when I have time.
Actually I just need somebody to tell me how to add this program as a daily prow job - noting more.

**Comment by grodrigues3** [comment]
I'll let @spxtr comment on that

**Comment by grodrigues3** [comment]
@lukaszgryglicki 

Does your algorithm not actually look at patterns? It looks like the [labels](https://github.com/lukaszgryglicki/test-infra/blob/260a807a3f33caa75deaf0e7e8910f7251b24ffb/labels.yaml) just has all the labels hardcoded.  

I want people to know that if someone tries to add a new `sig/somethingRandom` label to a repo they are aware that those are a protected pattern and for consistency, adding a sig/ label will result in their new sig label being added to all repos.

**Comment by lukaszgryglicki** [comment]
Yes it is not using any pattern. It uses hardcoded list of labels and sync them across the repos.
I think that functionality You suggest is something separate.
I was thinking that I need to take labels mentioned in doc from `kubernetes/kubernetes`, put them in labels.yaml (anyway - this file can be tweaked anytime, and next time my algorithm runs it will sync), create algorithm that sync this label set on all repos in given org.

**Comment by grodrigues3** [comment]
@lukaszgryglicki can you create the PR.  What are you blocked on?

**Comment by grodrigues3** [comment]
I think we still need a prow plugin to handle events where these protect patterns are created.  For consistency, we want exactly the same set of sig/, area/ etc labels.  If someone creates a new label with those prefixes we should handle that event.

**Comment by lukaszgryglicki** [comment]
I'm not really blocked. I need final suggestion how to "deploy" it.
PR created: https://github.com/kubernetes/test-infra/pull/2706

**Comment by lukaszgryglicki** [comment]
Well if we don't want to propagate across all repos, then we can call program with `-repos file_name.yaml` to restrict list of repositories.
So I think that now we only need to tweak `labels.yaml` and `repos.yaml` files and deploy it somehow.
We can start from just say 2-3 repositories and 4-5 most important labels and then update config yams files as needed.

**Comment by lukaszgryglicki** [comment]
@spxtr can You help me with this: https://github.com/kubernetes/test-infra/issues/2504#issuecomment-297106251
Specially:
```
It can just run as a daily job on prow, which means you can use our standard job control flow. The code should probably live outside of the prow folder. One example of such a job is the triage job, configured here, code here.
```

I don't know how to create this config entry for my new task:
1. `image: gcr.io/k8s-testimages/triage:v20170420-1ca6229c` - which image should I use. Or maybe I should create my own image with this tool installed? How can I compile it and put in this image?
2. My program will need some config files (like `labels.yaml` and possible `repos.yaml`), where should I put them? How can I call my program with some arguments (like paths of config files and paths to github oath token etc.)
3. `GOOGLE_APPLICATION_CREDENTIALS` -> `/etc/service-account/service-account.json` what does it mean? How can I put my own configuration for program. How can I get our prow's github secrets or configure it to use them?
4. You suggested to use other github token that prow's one. I see no problem - but somebody with access to kubernetes repo have to create that token.... we can use `prow` token IMHO - it have all what is needed for this program to work.
5. Should I configure github secrets in a manner similar to `https://github.com/kubernetes/test-infra/blob/master/triage/Dockerfile` ?

**Comment by spxtr** [comment]
> `image: gcr.io/k8s-testimages/triage:v20170420-1ca6229c` - which image should I use. Or maybe I should create my own image with this tool installed? How can I compile it and put in this image?

Create your own image! See the `Dockerfile` and `Makefile` in the triage folder I linked. Bonus points if you make a small image.

> My program will need some config files (like `labels.yaml` and possible `repos.yaml`), where should I put them? How can I call my program with some arguments (like paths of config files and paths to github oath token etc.)

The two config files have slightly different meaning. I think `labels.yaml` can either be pulled from GitHub each sync look or else it can be stored in a configmap on kubernetes. Either way is fine with me. IIUC, `repos.yaml` shouldn't be saved at all. Just regenerate it each time.

> `GOOGLE_APPLICATION_CREDENTIALS` -> `/etc/service-account/service-account.json` what does it mean? How can I put my own configuration for program. How can I get our prow's github secrets or configure it to use them?
>
> You suggested to use other github token that prow's one. I see no problem - but somebody with access to kubernetes repo have to create that token.... we can use `prow` token IMHO - it have all what is needed for this program to work.

GitHub credentials are stored on the prow cluster as k8s secrets, you can mount them in the prow config. I'm worried about sharing the same GitHub account due to the bursty nature of the labels syncer. If prow runs out of tokens then we'll really annoy developers. [Token usage](http://velodrome.k8s.io/dashboard/db/monitoring?panelId=1&fullscreen). Ideally we run the labels syncer as its own bot user. That might be more hassle than its worth, though, in which case I'll make a k8s-ci-robot token for this purpose. We will need to make extra sure that the labels syncer does not use more than one or two thousand tokens per hour at most.

> Should I configure github secrets in a manner similar to `https://github.com/kubernetes/test-infra/blob/master/triage/Dockerfile` ?

I don't think there are any secrets built-in to that image. Instead, we mount secrets when we run the pod.

**Comment by lukaszgryglicki** [comment]
1. OK so I'll create new image (smallest possible). Which one do You suggest as staring point? `Dockerfile`'s `FROM`? It should have golan installed imho to compile `label_sync`. Or maybe I should compile final binary and only put it in image? I don't know what are our standards there.

2. For `labels.yaml` and `repos.yaml` - to pull them from GitHub repo this needs to be merged first, right? Otherwise I can only pull them from my fork `lukaszgryglicki/test-infra` ... if they're stored in configmaps in kubernetes - how can I get them (we're using docker images here?)

3. I also think `repos.yaml` should not be used but.... people requested that sync should happen only on few repos (not all) at the beginning, so my idea is to run without sync first and generate `repos.yaml`. Then update this file with starting repos to sync and then instead of getting real org's repos, use that file instead. Also `labels.yaml` should have only few labels for a start.

4. Label syncer will be run once per day. First time it will use more points then in next days usually small number. I'm OK with creating new github token file specially for it, but I think I cannot do it - have no permission. Who can do it? And then give me some instructions how to mount that key in my own image? I'm even preferring new token for it - just to be sure I'm not disturbing normal k8s bot(s).

Thanks for answers.... I just have first `Dockerfile` that generates image which compiles my label_sync and displays `-help` output .... just for testing.

**Comment by spiffxp** [comment]
we lack documentation on what "standard" means, ref kubernetes/kubernetes #34255

kubernetes-dev thread: https://groups.google.com/d/msg/kubernetes-dev/0rTJ6wjcitg/VNH8fKArAAAJ

**Comment by lukaszgryglicki** [comment]
I did some updates, please take a look - this may be close to the finish, but I need feedback about Dockerfile, image and pod deployment yaml.

**Comment by kargakis** [comment]
Fixed by https://github.com/kubernetes/test-infra/pull/2706. Spawn new issues for follow-up work.
